File: services/geospatial-intelligence-service/src/geospatial_base/handlers/geocode.rs
use axum::{Json, extract::State};

use crate::{
    AppState,
    domain::geocoding,
    handlers::{ServiceResult, bad_request},
    models::spatial_index::{GeocodeRequest, GeocodeResponse, ReverseGeocodeRequest},
};

pub async fn forward_geocode(
    State(_state): State<AppState>,
    Json(request): Json<GeocodeRequest>,
) -> ServiceResult<GeocodeResponse> {
    if request.address.trim().is_empty() {
        return Err(bad_request("address is required"));
    }
    Ok(Json(geocoding::forward(&request.address)))
}

pub async fn reverse_geocode(
    State(_state): State<AppState>,
    Json(request): Json<ReverseGeocodeRequest>,
) -> ServiceResult<GeocodeResponse> {
    Ok(Json(geocoding::reverse(request.coordinate)))
}

File: services/ontology-exploratory-analysis-service/internal/handlers/geospatial/geocode.go
// Geocoding handlers ported 1:1 from
// services/ontology-exploratory-analysis-service/src/geospatial/geospatial_base/handlers/geocode.rs.
// Forward and reverse lookups delegate to the in-process gazetteer in
// internal/domain/geocoding (no DB / no external geocoder).

package geospatial

import (
	"encoding/json"
	"net/http"
	"strings"

	"github.com/openfoundry/openfoundry-go/services/ontology-exploratory-analysis-service/internal/domain/geocoding"
	"github.com/openfoundry/openfoundry-go/services/ontology-exploratory-analysis-service/internal/models"
)

// ForwardGeocode mirrors `handlers::geocode::forward_geocode`.
func (s *AppState) ForwardGeocode(w http.ResponseWriter, r *http.Request) {
	var req models.GeocodeRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if strings.TrimSpace(req.Address) == "" {
		writeError(w, http.StatusBadRequest, "address is required")
		return
	}
	writeJSON(w, http.StatusOK, geocoding.Forward(req.Address))
}

// ReverseGeocode mirrors `handlers::geocode::reverse_geocode`.
func (s *AppState) ReverseGeocode(w http.ResponseWriter, r *http.Request) {
	var req models.ReverseGeocodeRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	writeJSON(w, http.StatusOK, geocoding.Reverse(req.Coordinate))
}
